lib/src/modules/download_file.cc

```diff
@@ -113,31 +119,68 @@ namespace Modules {
     return ActionOutput { exit_code, std_out, std_err };
   }
 
+  ActionResponse DownloadFile::failure_response(const ActionRequest& request,
+                                      const fs::path& results_dir,
+                                      const std::string& message)
+  {
+    LOG_ERROR(message);
+    ActionResponse fail_response { ModuleType::Internal, request };
+    processOutputAndUpdateMetadata(fail_response);
+    fail_response.output = write_download_results(results_dir, EXIT_FAILURE, "", message);
+    return fail_response;
+  }
+
   // DownloadFile overrides callAction from the base BoltModule class since there's no need to run
   // any commands with DownloadFile. CallAction will simply download the file and return a result
   // based on if the download succeeded or failed.
   ActionResponse DownloadFile::callAction(const ActionRequest& request)
   {
     auto file_params = request.params();
     auto files = file_params.get<std::vector<lth_jc::JsonContainer>>("files");
-    auto destination = fs::path(files[0].get<std::string>("destination"));
     const fs::path& results_dir = request.resultsDir();
 
     ActionResponse response { ModuleType::Internal, request };
     lth_jc::JsonContainer result;
-    try {
-      Util::downloadFileFromMaster(master_uris_,
-                    file_download_connect_timeout_,
-                    file_download_timeout_,
-                    client_,
-                    module_cache_dir_->createCacheDir(files[0].get<std::string>("sha256")),
-                    destination,
-                    files[0]);
-      response.output = write_download_results(results_dir, EXIT_SUCCESS, "downloaded", "");
-    } catch (Module::ProcessingError& e) {
-      LOG_ERROR("Failed to download {1}; {2}", destination, e.what());
-      response.output = write_download_results(results_dir, EXIT_FAILURE, "", lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+    for (auto this_file : files) {
+      auto destination = fs::path(this_file.get<std::string>("destination"));
+      auto content_type = this_file.get<std::string>("content_type");
+      if (content_type == "file") {
```

Comment:
`content_type` seems a little strange as the name of the parameter, because it makes me think about the actual contents of the file/symlink/directory rather than the type of thing it is - I think `file_type` or `kind` or something like that might work.

But do we have to add a type param at all? I'm wondering if we could use a symlink if the `link_source` param is set, and otherwise use the same strategy as before?

(If this is here and called content_type to match some other code somewhere else, ignore me)

Reply:
Oh, nevermind about the second part, I was looking at the changeset as a whole, I didn't realize we kind of need the new parameter to deal with directories

Reply:
I had this var named "file_type" before, but Jonathan pointed out that it made things confusing to talk about (i.e. "file_type":"file" is awkward language to use).

I'll change it to `kind` though. I think that's probably a good idea.